File: kairospy/infrastructure/integrations/services/gateways/binance/earn.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import datetime, timezone
from decimal import Decimal

from kairospy.application.usecases.earn.domain import EarnPosition, EarnProduct, EarnRedeemRequest, EarnReward, EarnSubscribeRequest
from kairospy.infrastructure.integrations.application.connections import IntegrationConnectionSpec
from kairospy.infrastructure.integrations.domain import AccessScope, ProductFamily, TransportKind
from kairospy.infrastructure.integrations.services.connections.connection import Connection
from kairospy.infrastructure.integrations.services.gateways.binance.spot.client import BinanceSpotRestClient


class BinanceSimpleEarnConnection(Connection):
    def __init__(self, spec: IntegrationConnectionSpec) -> None:
        self.client = BinanceSpotRestClient(credential_id=spec.credential.id if spec.credential else None)
        super().__init__(spec, components=())

    def products(self, *, asset: str | None = None, product_type: str | None = None) -> Sequence[EarnProduct]:
        payload = self.client.get("/sapi/v1/simple-earn/products", params={"asset": asset, "productType": product_type}, signed=True)
        rows = payload.get("rows", ()) if isinstance(payload, Mapping) else ()
        return tuple(_product(row) for row in rows if isinstance(row, Mapping))

    def positions(self, *, asset: str | None = None) -> Sequence[EarnPosition]:
        payload = self.client.get("/sapi/v1/simple-earn/positions", params={"asset": asset}, signed=True)
        rows = payload.get("rows", ()) if isinstance(payload, Mapping) else ()
        return tuple(_position(row) for row in rows if isinstance(row, Mapping))

    def rewards(self, *, asset: str | None = None) -> Sequence[EarnReward]:
        payload = self.client.get("/sapi/v1/simple-earn/rewardsRecord", params={"asset": asset}, signed=True)
        rows = payload.get("rows", ()) if isinstance(payload, Mapping) else ()
        return tuple(_reward(row) for row in rows if isinstance(row, Mapping))

    def subscribe(self, request: EarnSubscribeRequest) -> object:
        return self.client.post("/sapi/v1/simple-earn/flexible/subscribe", params={"productId": request.product_id, "amount": str(request.amount)}, signed=True)

    def redeem(self, request: EarnRedeemRequest) -> object:
        return self.client.post("/sapi/v1/simple-earn/flexible/redeem", params={"productId": request.product_id, "amount": None if request.amount is None else str(request.amount), "destAccount": request.dest_account}, signed=True)


class BinanceSimpleEarnGateway:
    def open(self, spec: IntegrationConnectionSpec) -> BinanceSimpleEarnConnection:
        if spec.product is not ProductFamily.EARN or spec.access is not AccessScope.PRIVATE or spec.transport is not TransportKind.REST:
            raise ValueError("Binance Simple Earn gateway received an incompatible connection spec")
        return BinanceSimpleEarnConnection(spec)


def _product(row: Mapping[str, object]) -> EarnProduct:
    return EarnProduct(str(row.get("productId") or ""), str(row.get("asset") or ""), str(row.get("productType") or ""), _decimal(row.get("latestAnnualPercentageRate")), _decimal(row.get("minAmount")), _decimal(row.get("maxAmount")), str(row.get("status") or "unknown"))


def _position(row: Mapping[str, object]) -> EarnPosition:
    return EarnPosition(str(row.get("productId") or ""), str(row.get("asset") or ""), _decimal(row.get("totalAmount")), _decimal(row.get("latestAnnualPercentageRate")), str(row.get("status") or "unknown"), _time(row.get("updateTime")))


def _reward(row: Mapping[str, object]) -> EarnReward:
    return EarnReward(str(row.get("asset") or ""), _decimal(row.get("rewardsAmount")), str(row.get("productId")) if row.get("productId") is not None else None, _time(row.get("time")))


def _decimal(value: object) -> Decimal:
    try:
        return Decimal(str(value or "0"))
    except (TypeError, ValueError):
        return Decimal("0")


def _time(value: object) -> datetime | None:
    try:
        return None if value is None else datetime.fromtimestamp(float(value) / 1000, tz=timezone.utc)
    except (TypeError, ValueError):
        return None


__all__ = ["BinanceSimpleEarnConnection", "BinanceSimpleEarnGateway"]

File: crates/kairos-integration/src/services/gateways/binance/earn.rs
//! Binance Simple Earn capability.

use std::collections::BTreeMap;

use serde_json::Value;

use crate::application::{
    Connection, ConnectionSpec, EarnActionResult, EarnConnection, EarnPosition, EarnProduct,
    EarnProductType, EarnRedeemRequest, EarnReward, EarnSubscribeRequest,
};
use crate::domain::{AccessScope, IntegrationCapability, ProductFamily, TransportKind};
use crate::services::connections::ManagedConnection;

use super::spot::account::BinanceSpotAccountClient;

pub struct BinanceSimpleEarnConnection {
    connection: ManagedConnection,
    client: BinanceSpotAccountClient,
}

impl BinanceSimpleEarnConnection {
    pub fn new(
        api_key: impl Into<String>,
        secret: impl Into<String>,
        base_url: impl Into<String>,
    ) -> Result<Self, String> {
        let client = BinanceSpotAccountClient::new(api_key, secret, base_url)
            .map_err(|error| error.to_string())?;
        let spec = ConnectionSpec {
            connection_id: "account.binance.earn.rest".into(),
            provider: "binance".into(),
            product: Some(ProductFamily::Earn),
            access: AccessScope::Private,
            transport: TransportKind::Rest,
            capability: IntegrationCapability::Earn,
            credential_id: Some("binance".into()),
            asset_type: None,
        };
        Ok(Self {
            connection: ManagedConnection::new(spec, Vec::new())?,
            client,
        })
    }

    fn query(&self, path: &str, params: BTreeMap<String, String>) -> Result<Value, String> {
        self.client
            .signed_get(path, params)
            .map_err(|error| error.to_string())
    }

    fn action(
        &self,
        path: &str,
        params: BTreeMap<String, String>,
    ) -> Result<EarnActionResult, String> {
        let payload = self
            .client
            .signed_post(path, params)
            .map_err(|error| error.to_string())?;
        let success = payload
            .get("success")
            .and_then(Value::as_bool)
            .unwrap_or(true);
        Ok(EarnActionResult {
            accepted: success,
            action_id: payload
                .get("positionId")
                .or_else(|| payload.get("purchaseId"))
                .and_then(value_string),
            status: if success { "accepted" } else { "rejected" }.into(),
            reason: payload
                .get("msg")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .into(),
        })
    }
}

impl Connection for BinanceSimpleEarnConnection {
    fn identity(&self) -> &crate::domain::ConnectionIdentity {
        self.connection.identity()
    }
    fn state(&self) -> &crate::domain::ConnectionState {
        self.connection.state()
    }
    fn start(&mut self) -> Result<(), String> {
        self.connection.start()
    }
    fn stop(&mut self) -> Result<(), String> {
        self.connection.stop()
    }
    fn reconnect(&mut self) -> Result<(), String> {
        self.connection.reconnect()
    }
    fn health(&self) -> crate::domain::ConnectionHealth {
        self.connection.health()
    }
}

impl EarnConnection for BinanceSimpleEarnConnection {
    fn products(
        &mut self,
        asset: Option<&str>,
        product_type: Option<EarnProductType>,
    ) -> Result<Vec<EarnProduct>, String> {
        self.start()?;
        let mut params = BTreeMap::new();
        if let Some(asset) = asset {
            params.insert("asset".into(), asset.into());
        }
        if let Some(product_type) = product_type {
            params.insert(
                "productType".into(),
                product_type_name(product_type).to_ascii_uppercase(),
            );
        }
        let payload = self.query("/sapi/v1/simple-earn/products", params)?;
        Ok(payload
            .get("rows")
            .and_then(Value::as_array)
            .map(|rows| rows.iter().filter_map(normalize_product).collect())
            .unwrap_or_default())
    }

    fn positions(&mut self, asset: Option<&str>) -> Result<Vec<EarnPosition>, String> {
        self.start()?;
        let mut params = BTreeMap::new();
        if let Some(asset) = asset {
            params.insert("asset".into(), asset.into());
        }
        let payload = self.query("/sapi/v1/simple-earn/positions", params)?;
        Ok(payload
            .get("rows")
            .and_then(Value::as_array)
            .map(|rows| rows.iter().filter_map(normalize_position).collect())
            .unwrap_or_default())
    }

    fn rewards(&mut self, asset: Option<&str>) -> Result<Vec<EarnReward>, String> {
        self.start()?;
        let mut params = BTreeMap::new();
        if let Some(asset) = asset {
            params.insert("asset".into(), asset.into());
        }
        let payload = self.query("/sapi/v1/simple-earn/rewardsRecord", params)?;
        Ok(payload
            .get("rows")
            .and_then(Value::as_array)
            .map(|rows| rows.iter().filter_map(normalize_reward).collect())
            .unwrap_or_default())
    }

    fn subscribe(&mut self, request: &EarnSubscribeRequest) -> Result<EarnActionResult, String> {
        self.start()?;
        let path = match request.product_type {
            EarnProductType::Locked => "/sapi/v1/simple-earn/locked/subscribe",
            EarnProductType::Flexible => "/sapi/v1/simple-earn/flexible/subscribe",
        };
        let mut params = BTreeMap::from([
            (String::from("productId"), request.product_id.clone()),
            (String::from("amount"), request.amount.clone()),
        ]);
        if let Some(auto_renew) = request.auto_renew {
            params.insert("autoSubscribe".into(), auto_renew.to_string());
        }
        self.action(path, params)
    }

    fn redeem(&mut self, request: &EarnRedeemRequest) -> Result<EarnActionResult, String> {
        self.start()?;
        let path = match request.product_type {
            EarnProductType::Locked => "/sapi/v1/simple-earn/locked/redeem",
            EarnProductType::Flexible => "/sapi/v1/simple-earn/flexible/redeem",
        };
        let mut params = BTreeMap::from([(String::from("productId"), request.product_id.clone())]);
        if let Some(amount) = &request.amount {
            params.insert("amount".into(), amount.clone());
        }
        if let Some(destination) = &request.destination_account {
            params.insert("destAccount".into(), destination.clone());
        }
        self.action(path, params)
    }
}

fn normalize_product(value: &Value) -> Option<EarnProduct> {
    Some(EarnProduct {
        product_id: value.get("productId")?.as_str()?.into(),
        asset: value.get("asset")?.as_str()?.into(),
        product_type: product_type(
            value
                .get("productType")
                .and_then(Value::as_str)
                .unwrap_or("flexible"),
        ),
        annual_rate: value
            .get("latestAnnualPercentageRate")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        min_amount: value
            .get("minAmount")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        max_amount: value
            .get("maxAmount")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        status: value
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or("unknown")
            .into(),
        duration_days: value
            .get("duration")
            .and_then(Value::as_u64)
            .map(|value| value as u32),
    })
}

fn normalize_position(value: &Value) -> Option<EarnPosition> {
    Some(EarnPosition {
        product_id: value.get("productId")?.as_str()?.into(),
        asset: value.get("asset")?.as_str()?.into(),
        amount: value
            .get("totalAmount")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        rewards: value
            .get("totalRewards")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        annual_rate: value
            .get("latestAnnualPercentageRate")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        status: value
            .get("status")
            .and_then(Value::as_str)
            .unwrap_or("unknown")
            .into(),
        updated_at_unix_millis: value.get("updateTime").and_then(Value::as_u64),
    })
}

fn normalize_reward(value: &Value) -> Option<EarnReward> {
    Some(EarnReward {
        asset: value.get("asset")?.as_str()?.into(),
        amount: value
            .get("rewardsAmount")
            .map(value_string)
            .unwrap_or_default()
            .unwrap_or_default(),
        product_id: value
            .get("productId")
            .and_then(Value::as_str)
            .map(str::to_owned),
        occurred_at_unix_millis: value.get("time").and_then(Value::as_u64),
    })
}

fn product_type(value: &str) -> EarnProductType {
    if value.eq_ignore_ascii_case("locked") {
        EarnProductType::Locked
    } else {
        EarnProductType::Flexible
    }
}
fn product_type_name(value: EarnProductType) -> &'static str {
    match value {
        EarnProductType::Locked => "locked",
        EarnProductType::Flexible => "flexible",
    }
}
fn value_string(value: &Value) -> Option<String> {
    Some(
        value
            .as_str()
            .map(str::to_owned)
            .unwrap_or_else(|| value.to_string()),
    )
}
